### PR TITLE
Adding StepDefinitions View 

### PR DESCRIPTION
**Pull request by jeremiemarc**
Hello Pierre, 

I've added a view to visualise the step definitions used during a test run.
Can you have a look at it please?

Let me know if there's any modification you'd like to see ;-)

image
image

Jérémi

### CODE REVIEW

**Comment by pgentile** [comment]
Update the CHANGELOG.md file

File: zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/views/GroupedStepsListItemView.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+import io.zucchiniui.backend.scenario.domain.Step;
+import io.zucchiniui.backend.shared.domain.Location;
+
+import java.util.Set;
+
+public class GroupedStepsListItemView {
+
+    private Location stepDefinitionLocation;
+    private Set<Step> occurrences;
```
Comment (pgentile) [inline zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/views/GroupedStepsListItemView.java]: A Occurence object would be better, with only properties info and status from the step.

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+  return formattedText;
+}
+
+function formatStepDefinitionVariants(occurrences) {
+
+  const blockStyle = { float: 'left' };
+  const keywordStyle = 'step-definition-keyword step-definition-variant';
+  const argumentStyle = 'step-definition-argument step-definition-variant';
+  const definition = occurrences[0].info.name;
+  const keywords = [...new Set(occurrences.map(occurrence => { return occurrence.info.keyword }))];
```
Comment (pgentile) [inline zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: Use Array.from for Set conversion : https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    formattedText.push(definition.substr(startIdx));
+  } else {
+    formattedText.push(definition)
+  }
+  return formattedText;
+}
+
+function formatStepDefinitionVariants(occurrences) {
+
+  const blockStyle = { float: 'left' };
```
Comment (pgentile) [inline zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: Put style in the styles.less file

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+function formatVariantsBlock(variants, variantStyle) {
+  return variants.map(variant => {
+    return <StepDefinitionsTableHighlightedTerm key={variant} text={variant} styleClass={variantStyle} />
+  });
+}
+
+function formatStepDefinition(stepDefinition) {
+
+  const keyword = stepDefinition.info.keyword;
+  const definition = truncate(stepDefinition.info.name, { 'length': 300 });
```
Comment (pgentile) [inline zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: is truncate really needed ?

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+          }
+        }
+      })
+    });
+  }
+  return groupedArgs;
+}
+
+function getClosestOffset(knownOffsets, currentOffset) {
+  // Sometimes the offset returned by cucumber is slightly shifted due to non-capturing groups in regexes
```
Comment (pgentile) [inline zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: Regex are aweful :)

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    }
+    this.props.onClose();
+  };
+
+  render() {
+    const { show, occurrences, location } = this.props;
+    return (
+      <Modal bsSize='large' dialogClassName='modal-dialog' show={show} onHide={this.onCloseClick}>
+        <Modal.Header closeButton>
+          <Modal.Title>{'Variantes Connues'} <span className='step-definition-location'>{location}</span></Modal.Title>
```
Comment (pgentile) [inline zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: `{'Variantes Connues'}` -> `Variantes Connues`

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+function formatStepDefinition(stepDefinition) {
+
+  const keyword = stepDefinition.info.keyword;
+  const definition = truncate(stepDefinition.info.name, { 'length': 300 });
+  const args = stepDefinition.info.arguments;
+
+  const formattedText = [];
+  formattedText.push(<span key={Math.random()}><StepDefinitionsTableHighlightedTerm text={keyword} styleClass='step-definition-keyword' />&nbsp;</span>);
```
Comment (pgentile) [inline zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: Never use a random for the key attribute in React. Use an index variable for all keys.

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+  const blockStyle = { float: 'left' };
+  const keywordStyle = 'step-definition-keyword step-definition-variant';
+  const argumentStyle = 'step-definition-argument step-definition-variant';
+  const definition = occurrences[0].info.name;
+  const keywords = [...new Set(occurrences.map(occurrence => { return occurrence.info.keyword }))];
+  const groupedArgs = groupDistinctArgumentsByOffset(occurrences);
+
+  const formattedText = [];
+  formattedText.push(<div style={blockStyle} key={Math.random()}>{formatVariantsBlock(keywords, keywordStyle)}</div>);
```
Comment (pgentile) [inline zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: Never use a random for the key attribute in React. Use an index variable for all keys.

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+
+  let startIdx = 0;
+  for (const offset in groupedArgs) {
+    const position = parseInt(offset);
+    if (groupedArgs.hasOwnProperty(position)) {
+      const variants = groupedArgs[position];
+      formattedText.push(<div style={blockStyle} key={Math.random()}>{definition.substr(startIdx, position - startIdx)}</div>);
```
Comment (pgentile) [inline zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: Never use a random for the key attribute in React. Use an index variable for all keys.

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+  render() {
+    const { occurrence } = this.props;
+    const highlightedText = formatStepDefinition(occurrence);
+    return (
+      <span>{highlightedText.map((text, idx) => <span key={idx}>{text}</span>)}</span>
+    );
+  }
+}
+
+class StepDefinitionsTableHighlightedTerm extends React.Component {
```
Comment (pgentile) [inline zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: Remove this component, it is not useful for now.

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+
+  render() {
+    const { text, styleClass } = this.props;
+    return (
+      <span className={styleClass}>{text}</span>
+    );
+  }
+}
+
+class StepDefinitionsVariantsDialog extends React.PureComponent {
```
Comment (pgentile) [inline zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: The StepDefinitionsVariantsDialog should be in its own file (with its functions).

File: zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/views/ScenarioViewAccess.java [hunk 12]
```diff
@@ -190,4 +191,38 @@ public ScenarioStats getStats(final Consumer<ScenarioQuery> preparator) {
...
 
+    public List<GroupedStepsListItemView> getStepDefinitions(final Consumer<ScenarioQuery> preparator) {
+
+        final Query<Scenario> query = scenarioDAO.prepareTypedQuery(preparator).project("steps", true);
+        List<Step> steps = MorphiaUtils.streamQuery(query)
+            .flatMap(scenario -> scenario.getSteps().stream())
+            .collect(Collectors.toList());
+
+        List<GroupedStepsListItemView> groupedSteps = new ArrayList<>();
+        steps.stream().filter(step -> step.getStepDefinitionLocation() != null).
```
Comment (pgentile) [inline zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/views/ScenarioViewAccess.java]: Last dot with the `.forEach`

Comment (pgentile) [reply zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/views/ScenarioViewAccess.java]: You can apply the filter between the flatMap and collect operations of the `MorphiaUtils.streamQuery(query)` stream.

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+      startIdx = arg.offset + arg.value.length;
+    });
+    formattedText.push(definition.substr(startIdx));
+  } else {
+    formattedText.push(definition)
+  }
+  return formattedText;
+}
+
+function formatStepDefinitionVariants(occurrences) {
```
Comment (pgentile) [inline zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: You should generate HTML directly and not use React. Look at the ElementInfo component for that.

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+
+  const keyword = stepDefinition.info.keyword;
+  const definition = truncate(stepDefinition.info.name, { 'length': 300 });
+  const args = stepDefinition.info.arguments;
+
+  const formattedText = [];
+  formattedText.push(<span key={Math.random()}><StepDefinitionsTableHighlightedTerm text={keyword} styleClass='step-definition-keyword' />&nbsp;</span>);
+
+  let startIdx = 0;
+  if (args !== undefined) {
```
Comment (pgentile) [inline zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: The following looks like the `tokenizeFromInfo` function.
Maybe you can simplify the grunt work with `tokenizeFromInfo` (but you may need to add the offset of each token, in the `tokenizeFromInfo` function).

File: zucchini-ui-backend/src/main/java/io/zucchiniui/backend/reportconverter/converter/ReportScenarioConverter.java [hunk 15]
```diff
@@ -117,6 +119,7 @@ private static void buildStep(final ReportStep reportStep, final StepBuilder ste
             .withErrorMessage(reportStep.getResult().getErrorMessage())
             .withStatus(convertStepStatus(reportStep.getResult().getStatus()))
             .withInfo(stepInfo)
+            .withStepDefinitionLocation(convertLocation(reportStep.getMatch().getLocation(), ":"))
```
[the remark this answers is not in the document]
Comment (jeremiemarc) [reply zucchini-ui-backend/src/main/java/io/zucchiniui/backend/reportconverter/converter/ReportScenarioConverter.java]: OK! The attribute & corresponding method have been modified

File: zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/domain/Step.java [hunk 16]
```diff
@@ -23,6 +24,8 @@
 
     private List<Attachment> attachments = new ArrayList<>();
 
+    private Location stepDefinitionLocation;
```
[the remark this answers is not in the document]
Comment (jeremiemarc) [reply zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/domain/Step.java]: Ok

File: zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/domain/StepBuilder.java [hunk 17]
```diff
@@ -1,13 +1,16 @@
...
 import io.zucchiniui.backend.shared.domain.BasicInfo;
+import io.zucchiniui.backend.shared.domain.Location;
 
 import java.util.List;
 
 public class StepBuilder {
 
     private BasicInfo info;
 
+    private Location stepDefinitionLocation;
```
[the remark this answers is not in the document]
Comment (jeremiemarc) [reply zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/domain/StepBuilder.java]: OK

File: zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/rest/ScenarioResource.java [hunk 18]
```diff
@@ -121,6 +122,19 @@ public ScenarioStats getStats(@BeanParam final GetScenariiRequestParams requestP
...
     }
 
+    @GET
+    @Path("stepDefinitions")
+    public List<GroupedStepsListItemView> getGroupedStepDefinitions(@BeanParam final GetScenariiRequestParams requestParams) {
+        return scenarioViewAccess.getStepDefinitions(q -> {
+            if (!Strings.isNullOrEmpty(requestParams.getTestRunId())) {
+                q.withTestRunId(requestParams.getTestRunId());
+            }
+            if (!Strings.isNullOrEmpty(requestParams.getFeatureId())) {
```
[the remark this answers is not in the document]
Comment (jeremiemarc) [reply zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/rest/ScenarioResource.java]: ok

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx (hunk 3 above)
Comment (jeremiemarc) [reply zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: Done

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx (hunk 5 above)
Comment (jeremiemarc) [reply zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: ;-)

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx (hunk 8 above)
Comment (jeremiemarc) [reply zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: I removed the random and used a unique combination of the object's attributes

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx (hunk 9 above)
Comment (jeremiemarc) [reply zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: OK

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx (hunk 10 above)
Comment (jeremiemarc) [reply zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: I'm not sure that I understand your request ;-). I use this component to display both keywords & variants (and make them stand out)

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx (hunk 11 above)
Comment (jeremiemarc) [reply zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: OK. Extracted to its file

File: zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx (hunk 2 above)
Comment (jeremiemarc) [reply zucchini-ui-frontend/src/stepDefinitions/components/StepDefinitionsTable.jsx]: ok

File: zucchini-ui-frontend/src/styles/main.less [hunk 19]
```diff
@@ -148,7 +148,29 @@
   opacity: 0;
 }
 
-.error-message-modal-dialog {
-  max-width: 100%;
+.modal-dialog {
```
[the remark this answers is not in the document]
Comment (jeremiemarc) [reply zucchini-ui-frontend/src/styles/main.less]: Ooops.. Did not notice that I was overriding that! ;-)
Fixed

**Comment by jeremiemarc** [comment]
Hello @pgentile ,
I've taken your change requests into account

**Pull request merged**